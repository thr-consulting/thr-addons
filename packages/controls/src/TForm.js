// @flow

import React, {Component} from 'react';
import isEmpty from 'lodash/isEmpty';
import {Formik} from 'formik';
import {Message} from 'semantic-ui-react';

type Props = {
	/** Render prop just like Formik's with additional arguments passed */
	render: any,
	/** The number of fields in the form. TForm can't automatically get this number. */
	numFields: number,
	/** Used to pass in any GraphQL errors. */
	errors?: any,
	/** Called when the form submits. */
	onSubmit: data => void,
};

/**
 * Extends Formik to provide Semantic UI error and warning messages and field errors.
 * @class
 * @property {function} render - Render prop just like Formik's with additional arguments passed
 * @property {number} numFields - The number of fields in the form. TForm can't automatically get this number.
 * @property errors - Used to pass in any GraphQL errors.
 * @property {function} onSubmit - Called when the form submits.
 */
export default class TForm extends Component<Props> {
	static displayName = 'TForm';

	props: Props;

	renderForm = args => {
		const {errors: warnings, touched} = args;
		const {errors, numFields, render} = this.props;

		return render({
			renderErrors() {
				if (!errors) return null;
				const {graphQLErrors} = errors;
				return (
					<Message error>
						<Message.Header>Error</Message.Header>
						{graphQLErrors.map(err => (
							<p key={err.message}>{err.message}</p>
						))}
					</Message>
				);
			},
			renderWarnings() {
				if (isEmpty(warnings)) return null;
				return (
					<Message warning>
						<Message.Header>Please check the following fields:</Message.Header>
						<ul>
							{Object.keys(warnings).map(warning => (
								<li key={warning}>{warnings[warning]}</li>
							))}
						</ul>
					</Message>
				);
			},
			hasErrors() {
				return !!errors;
			},
			hasWarnings() {
				return Object.keys(touched).length === numFields && !isEmpty(warnings);
			},
			fieldError(fieldName) {
				return !!(touched[fieldName] && warnings[fieldName]);
			},
			...args,
		});
	};

	render() {
		const {render, numFields, ...rest} = this.props;
		return (
			<Formik render={this.renderForm} {...rest}/>
		);
	}
}
