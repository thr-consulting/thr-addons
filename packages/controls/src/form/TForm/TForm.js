// @flow

import React, {Component} from 'react';
import isEmpty from 'lodash/isEmpty';
import isString from 'lodash/isString';
import {Formik} from 'formik';
import {Message} from 'semantic-ui-react';

type Props = {
	/** Render prop just like Formik's with additional arguments passed */
	render: any,
	/** The number of fields in the form. TForm can't automatically get this number. */
	numFields: number,
	/** Used to pass in any GraphQL errors. */
	errors?: any,
	/** Called when the form submits. */
	onSubmit: data => void,
	/** Called when the TForm is mounted so you can get access to the submitForm function. */
	getSubmitFn: fn => void,
};

/**
 * Extends Formik to provide Semantic UI error and warning messages and field errors.
 * @class
 * @property {function} render - Render prop just like Formik's with additional arguments passed
 * @property {number} numFields - The number of fields in the form. TForm can't automatically get this number.
 * @property errors - Used to pass in any GraphQL errors.
 * @property {function} onSubmit - Called when the form submits.
 * @property {function} getSubmitFn - Called when the TForm is mounted so you can get access to the submitForm function.
 */
export default class TForm extends Component<Props> {
	static displayName = 'TForm';

	componentDidMount() {
		if (this.props.getSubmitFn) this.props.getSubmitFn(this._submitForm);
	}


	renderForm = args => {
		const {errors: warnings, touched, handleChange, setFieldValue, submitForm, ...rest} = args;
		const {errors, numFields, render} = this.props;

		this._submitForm = submitForm;

		return render({
			renderErrors() {
				if (!errors) return null;
				const {graphQLErrors} = errors;
				return (
					<Message error>
						<Message.Header>Error</Message.Header>
						{graphQLErrors.map(err => (
							<p key={err.message}>{err.message}</p>
						))}
					</Message>
				);
			},
			renderWarnings() {
				if (isEmpty(warnings)) return null;
				return (
					<Message warning>
						<Message.Header>Some fields are not complete:</Message.Header>
						<ul>
							{Object.keys(warnings).map(warning => (
								<li key={warning}>{warnings[warning]}</li>
							))}
						</ul>
					</Message>
				);
			},
			hasErrors() {
				return !!errors;
			},
			hasWarnings() {
				return Object.keys(touched).length === numFields && !isEmpty(warnings);
			},
			fieldError(fieldName) {
				return !!(touched[fieldName] && warnings[fieldName]);
			},
			handleChange(evOrName) {
				if (evOrName.nativeEvent) return handleChange(evOrName);
				if (isString(evOrName)) {
					return val => {
						setFieldValue(evOrName, val);
					};
				}
				throw new Error('TForm expects handleChange to receive a SyntheticEvent or a string value of the field name');
			},
			errors: warnings,
			touched,
			setFieldValue,
			submitForm,
			...rest,
		});
	};

	render() {
		const {render, numFields, ...rest} = this.props;
		return (
			<Formik render={this.renderForm} {...rest}/>
		);
	}
}
