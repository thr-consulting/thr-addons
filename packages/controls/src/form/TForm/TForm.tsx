import debug from 'debug';
import React from 'react';
import {Formik, FormikConfig, FormikValues} from 'formik';
import {TFormInner, TFormProps} from './TFormInner';

const d = debug('thx.controls.TForm');

export default function TForm<Values extends FormikValues = FormikValues>(
	props: FormikConfig<Values> & TFormProps<Values>,
) {
	return (
		<Formik<Values> {...props}>
			{formikProps => {
				return <TFormInner<Values> formikProps={formikProps} tFormProps={props} children={props.children} />;
			}}
		</Formik>
	);
}
