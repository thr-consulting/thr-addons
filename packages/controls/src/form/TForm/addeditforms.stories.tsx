import {useArgs} from '@storybook/client-api';
import type {Meta} from '@storybook/react';
import debug from 'debug';
import {Form, Input} from 'semantic-ui-react';
import {number, object, string} from 'yup';
import type {TFormConfig, TFormProps} from './types';
import {useTForm} from './useTForm';

const d = debug('thx.controls.form.TForm.addeditforms.stories');

export default {
	title: 'Form/Add and Edit Forms',
	argTypes: {
		onSubmit: {type: 'function'},
		onChange: {type: 'function'},
		onValidate: {type: 'function'},
	},
} as Meta;

interface LockedEditForm {
	id: string;
	version: number;
}

function lockedEditFormValidation(a: any) {
	return a.clone().shape({
		id: string().required(),
		version: number().required(),
	});
}

/* ***************************************************************************** */

const addressFormValidation = object().shape({
	line1: string().required(),
	line2: string().notRequired(),
});
const addressEditFormValidation = lockedEditFormValidation(addressFormValidation);

interface AddressFormType {
	line1: string;
	line2?: string;
}

type AddressEditFormType = AddressFormType & LockedEditForm;

function AddressForm<T extends AddressFormType>(props: TFormProps<T>) {
	const {handleSubmit, formError, hasWarnings, values, handleChange, handleBlur, submitDisabled, isSubmitting} = props;

	return (
		<Form onSubmit={handleSubmit} error={formError} warning={hasWarnings}>
			<Form.Field width={6}>
				<label>Line 1</label>
				<Input name="line1" value={values.line1} onChange={handleChange} onBlur={handleBlur} />
			</Form.Field>
			<Form.Field width={6}>
				<label>Line 2</label>
				<Input name="line2" value={values.line2} onChange={handleChange} onBlur={handleBlur} />
			</Form.Field>
			<Form.Button disabled={submitDisabled} type="submit" loading={isSubmitting}>
				Submit
			</Form.Button>
		</Form>
	);
}

export function AddressCreateForm(args: {onSubmit: (value: any) => any}) {
	const [, updateArgs] = useArgs();

	const tform = useTForm<AddressFormType>({
		initialValues: {line1: '', line2: ''},
		validationSchema: addressFormValidation,
		onSubmit(value) {
			args.onSubmit && args.onSubmit(addressFormValidation.validateSync(value, {stripUnknown: true}));
		},
	});

	return <AddressForm {...tform} />;
}
AddressCreateForm.args = {
	enableReinitialize: false,
	loading: false,
};

export function AddressEditForm(args: TFormConfig<AddressEditFormType>) {
	const [, updateArgs] = useArgs();

	const tform = useTForm<AddressEditFormType>({
		...args,
		initialValues: {line1: '', line2: ''},
		validationSchema: addressEditFormValidation,
		async onSubmit(value, helpers) {
			await args.onSubmit(addressEditFormValidation.validateSync(value, {stripUnknown: true}), helpers);
		},
	});

	return <AddressForm {...tform} />;
}
AddressEditForm.args = {
	enableReinitialize: false,
	loading: false,
};
