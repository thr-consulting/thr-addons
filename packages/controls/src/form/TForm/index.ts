export {default} from './TForm';
export {TFormProps, TFormChildrenProps, TFormInner, TFormInnerProps} from './TFormInner';
