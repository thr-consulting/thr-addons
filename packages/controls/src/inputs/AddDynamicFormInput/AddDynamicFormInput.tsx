import {Label, Grid, Button, Form, FormInput} from 'semantic-ui-react';
import React, {Children, cloneElement, ReactElement} from 'react';
import debug from 'debug';
import {FieldArray} from 'formik';
import isEmpty from 'lodash/isEmpty';
import {SemanticWIDTHS} from 'semantic-ui-react/dist/commonjs/generic';

const d = debug('AddDynamicFormInputs');

interface Props {
	values: object[],
	setFieldValue: (position: string, value: {}) => void,
	name: string,
	label: string,
	disableAddName?: boolean,
	children: ReactElement[],
	fieldError: (name: string | string[]) => boolean,
	onBlur: (e: any) => void,
}

/**
 * Takes Form.Input as children
 * @param props
 * @constructor
 */
export default function AddDynamicFormInput(props: Props) {
	const handleChange = (e, child): void => {
		const {name, value} = child;
		props.setFieldValue(name, value);
	};

	const renderAddButton = (index, label, arrayHelpers) => (
		<Form.Input style={{marginTop: 25}}>
			<Label fluid="true" size="large">{label}</Label>
			<Button color="green" disabled={props.disableAddName} onClick={() => arrayHelpers.push({})} type="button">Add</Button>
		</Form.Input>
	);

	const renderGrid = (arrayHelpers, arrayObj = {}, rowIndex = 0, length = 0) => {
		const {children, label = 'label', fieldError, name, onBlur} = props;
		if (!children) throw new Error('There are no children defined for AddDynamicFormInput');

		return (
			<Grid stackable columns={children.length + 1 as SemanticWIDTHS} key={rowIndex}>
				<Grid.Row>
					<Grid.Column>
						{rowIndex < 1 ? renderAddButton(length, label, arrayHelpers) : null}
					</Grid.Column>
					{Children.map(children, (Child: ReactElement, columnIndex) => {
						const {type, props: childProps} = Child;
						if (type !== FormInput) return <Form.Input label="Invalid Input Provided" placeholder="Must be a Form.Input" error/>;
						// eslint-disable-next-line
						const {onChange, value, error, name: childName, ...rest} = childProps;

						// if there is more than one column and it is the last child element.
						const deleteAction = (
							(length > 1) && (columnIndex === children.length - 1))
							? {icon: 'delete', color: 'red', onClick: () => arrayHelpers.remove(rowIndex)}
							: null;

						const n = `${name}.${rowIndex.toString()}.${childName}`;

						const childElement = (
							cloneElement(Child, {
								name: n,
								index: rowIndex,
								error: fieldError(n),
								action: deleteAction,
								onChange: handleChange,
								value: arrayObj[childProps.name] || '',
								onBlur,
								...rest,
							}));

						return (<Grid.Column key={name}>{childElement}</Grid.Column>);
					})}
				</Grid.Row>
			</Grid>
		);
	};

	const renderGridWithValues = arrayHelpers => {
		const {values} = props;
		return values.map((arrayObj, index) => renderGrid(arrayHelpers, arrayObj, index, values.length));
	};

	return (
		<FieldArray
			name={props.name}
			render={arrayHelpers => (
				isEmpty(props.values) ? renderGrid(arrayHelpers) : renderGridWithValues(arrayHelpers)
			)}
		/>
	);
}
