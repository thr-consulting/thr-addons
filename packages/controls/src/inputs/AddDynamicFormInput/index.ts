export {AddDynamicFormInput, AddDynamicFormInputProps} from './AddDynamicFormInput';
