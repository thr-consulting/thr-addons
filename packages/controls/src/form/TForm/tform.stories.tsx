import {Button, Group, TextInput} from '@mantine/core';
/* eslint-disable react-hooks/rules-of-hooks */
import {useArgs} from '@storybook/client-api';
import type {Meta} from '@storybook/react';
import debug from 'debug';
import {useState} from 'react';
import {object, string} from 'yup';
import {storyDecorator} from '../../storyDecorator';
import {TForm} from './TForm';
import type {TFormProps} from './types';

const d = debug('thx.controls.form.TForm.tform.stories');

export default {
	title: 'Form/TForm',
	argTypes: {
		onSubmit: {type: 'function'},
		onChange: {type: 'function'},
		onValidate: {type: 'function'},
	},
	decorators: [
		storyDecorator,
		Story => (
			<div style={{width: 400, position: 'relative'}}>
				<Story />
			</div>
		),
	],
} as Meta;

// Yup Validation
const formValidation = object().shape({
	text: string().required('Text is required'),
	nested: object().shape({
		text: string().required('Nested text is required'),
	}),
	plainInput: string(),
});

interface FormType {
	text: string;
	nested: {
		text: string;
	};
	plainInput: string;
}

function MyForm(props: TFormProps<FormType>) {
	const {values, handleChange, handleSubmit, handleBlur, submitDisabled, isSubmitting, renderWarnings} = props;

	// error={formError} warning={hasWarnings}
	return (
		<form onSubmit={handleSubmit}>
			<TextInput name="text" label="Enter some text" value={values.text} onChange={handleChange} onBlur={handleBlur} />
			<TextInput name="nested.text" label="Enter some nested text" value={values.nested.text} onChange={handleChange} onBlur={handleBlur} />
			<TextInput name="plainInput" label="Plain Input" value={values.plainInput} onChange={handleChange} onBlur={handleBlur} />
			<Group position="right" mt="md">
				<Button disabled={submitDisabled} type="submit" loading={isSubmitting}>
					Submit
				</Button>
			</Group>
			{renderWarnings()}
		</form>
	);
}

export function Main(args: any) {
	const [, updateArgs] = useArgs();

	return (
		<TForm<FormType>
			{...args}
			initialValues={{text: '', nested: {text: ''}, plainInput: ''}}
			validationSchema={formValidation}
			onSubmit={async (data, helpers) => {
				args.onSubmit(data, helpers);
			}}
		>
			{props => <MyForm {...props} />}
		</TForm>
	);
}
Main.args = {
	enableReinitialize: false,
	loading: false,
};

/* **************************************************************************** */

interface ApolloError {
	message?: string;
	graphQLErrors?: {
		message?: string;
	}[];
}

const sampleGraphqlError: (ApolloError | undefined)[] = [
	undefined,
	{message: 'Sample Graphql Error 1'},
	{
		message: 'Sample Graphql Error 2',
		graphQLErrors: [{message: 'Graphql Sub Error 2-A'}, {message: 'Graphql Sub Error 2-B'}],
	},
];

export function WithError(args: any) {
	// const [, updateArgs] = useArgs();
	const [graphqlError, setGraphqlError] = useState(0);

	return (
		<>
			<div style={{width: 400, position: 'relative'}}>
				<TForm<FormType>
					{...args}
					error={sampleGraphqlError[graphqlError]}
					initialValues={{text: '', nested: {text: ''}, plainInput: ''}}
					validationSchema={formValidation}
					onSubmit={data => {
						return new Promise((resolve, reject) => {
							setTimeout(() => {
								if (graphqlError === 0) {
									args.onSubmit(data);
									resolve(data);
								} else {
									// updateArgs({error: sampleGraphqlError[graphqlError]});
									reject();
								}
							}, 1000);
						});
					}}
				>
					{props => <MyForm {...props} />}
				</TForm>
			</div>
			<Group position="left" mt="md">
				<Button
					color={graphqlError === 0 ? 'green' : 'blue'}
					onClick={() => {
						setGraphqlError(0);
						// updateArgs({error: undefined});
					}}
				>
					Clear GraphQL Error
				</Button>
				<Button
					color={graphqlError === 1 ? 'green' : 'blue'}
					onClick={() => {
						setGraphqlError(1);
						// updateArgs({error: undefined});
					}}
				>
					Simulate GraphQL Error 1
				</Button>
				<Button
					color={graphqlError === 2 ? 'green' : 'blue'}
					onClick={() => {
						setGraphqlError(2);
						// updateArgs({error: undefined});
					}}
				>
					Simulate GraphQL Error 2
				</Button>
			</Group>
		</>
	);
}
WithError.args = {
	...Main.args,
	error: undefined,
};

// export function UsingHooks(args) {
// 	const [, updateArgs] = useArgs();
//
// 	const props = useTForm<FormType>({
// 		...args,
// 		initialValues: {
// 			text: '',
// 			nested: {
// 				text: '',
// 			},
// 			plainInput: '',
// 		},
// 		validationSchema: formValidation,
// 		onSubmit: data => args.onSubmit(data),
// 	});
//
// 	return <MyForm {...props} />;
// }
//
// UsingHooks.args = {
// 	...Main.args,
// };
//
// export function OutsideSubmit(args) {
// 	const [, updateArgs] = useArgs();
// 	const formSubmitFn = useRef<() => Promise<void>>();
//
// 	return (
// 		<>
// 			<TForm<FormType>
// 				{...args}
// 				initialValues={{text: '', nested: {text: ''}, plainInput: ''}}
// 				validationSchema={formValidation}
// 				getSubmitFn={sub => (formSubmitFn.current = sub)}
// 				onSubmit={async (data, formikHelpers) => {
// 					args.onSubmit(data, formikHelpers);
// 				}}
// 			>
// 				{props => <MyForm {...props} />}
// 			</TForm>
// 			<Segment>
// 				<Button color="orange" onClick={() => formSubmitFn.current && formSubmitFn.current()}>
// 					Outside Submit
// 				</Button>
// 			</Segment>
// 		</>
// 	);
// }
//
// OutsideSubmit.args = {
// 	...Main.args,
// };
