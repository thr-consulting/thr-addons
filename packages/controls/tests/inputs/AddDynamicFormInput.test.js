import React from 'react';
import {mount} from 'enzyme';
import {Form} from 'semantic-ui-react';
import AddDynamicFormInput from '../../src/inputs/AddDynamicFormInput';
import TForm from '../../src/form/TForm';

describe('MaskedInput', () => {
	it('should render without throwing an error', () => {
		expect(mount(
			<TForm
				render={({setFieldValue, values, fieldError, handleBlur}) => (
					<AddDynamicFormInput
						name="addDynamicFormInputs"
						label="Name(s) of account holder(s):"
						setFieldValue={setFieldValue}
						values={values.addDynamicFormInputs}
						fieldError={fieldError}
						onBlur={handleBlur}
					>
						<Form.Input name="firstName" label="First Name"/>
						<Form.Input name="lastName" label="Last Name"/>
					</AddDynamicFormInput>
				)}
			/>
		)).toMatchSnapshot();
	});
});
