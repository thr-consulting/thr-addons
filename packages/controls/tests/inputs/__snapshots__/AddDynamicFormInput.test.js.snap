// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskedInput should render without throwing an error 1`] = `
<TForm
  render={[Function]}
>
  <Formik
    enableReinitialize={true}
    isInitialValid={false}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={true}
  >
    <RenderForm
      dirty={false}
      errors={Object {}}
      handleBlur={[Function]}
      handleChange={[Function]}
      handleReset={[Function]}
      handleSubmit={[Function]}
      initialValues={Object {}}
      isSubmitting={false}
      isValid={false}
      isValidating={false}
      registerField={[Function]}
      resetForm={[Function]}
      setError={[Function]}
      setErrors={[Function]}
      setFieldError={[Function]}
      setFieldTouched={[Function]}
      setFieldValue={[Function]}
      setFormikState={[Function]}
      setStatus={[Function]}
      setSubmitting={[Function]}
      setTouched={[Function]}
      setValues={[Function]}
      submitCount={0}
      submitForm={[Function]}
      tFormProps={
        Object {
          "render": [Function],
        }
      }
      touched={Object {}}
      unregisterField={[Function]}
      validateField={[Function]}
      validateForm={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
      values={Object {}}
    >
      <AddDynamicFormInput
        fieldError={[Function]}
        label="Name(s) of account holder(s):"
        name="addDynamicFormInputs"
        onBlur={[Function]}
        setFieldValue={[Function]}
      >
        <FormikConnect(FieldArrayInner)
          name="addDynamicFormInputs"
          render={[Function]}
        >
          <FieldArrayInner
            formik={
              Object {
                "dirty": false,
                "errors": Object {},
                "handleBlur": [Function],
                "handleChange": [Function],
                "handleReset": [Function],
                "handleSubmit": [Function],
                "initialValues": Object {},
                "isSubmitting": false,
                "isValid": false,
                "isValidating": false,
                "registerField": [Function],
                "resetForm": [Function],
                "setError": [Function],
                "setErrors": [Function],
                "setFieldError": [Function],
                "setFieldTouched": [Function],
                "setFieldValue": [Function],
                "setFormikState": [Function],
                "setStatus": [Function],
                "setSubmitting": [Function],
                "setTouched": [Function],
                "setValues": [Function],
                "status": undefined,
                "submitCount": 0,
                "submitForm": [Function],
                "touched": Object {},
                "unregisterField": [Function],
                "validate": undefined,
                "validateField": [Function],
                "validateForm": [Function],
                "validateOnBlur": true,
                "validateOnChange": true,
                "validationSchema": undefined,
                "values": Object {},
              }
            }
            name="addDynamicFormInputs"
            render={[Function]}
            validateOnChange={true}
          >
            <Grid
              columns="equal"
              key="0"
              stackable={true}
            >
              <div
                className="ui stackable equal width grid"
              >
                <GridRow>
                  <div
                    className="row"
                  >
                    <GridColumn>
                      <div
                        className="column"
                      >
                        <FormInput
                          as={[Function]}
                          control={[Function]}
                          style={
                            Object {
                              "marginTop": 25,
                            }
                          }
                        >
                          <FormField
                            control={[Function]}
                            style={
                              Object {
                                "marginTop": 25,
                              }
                            }
                          >
                            <div
                              className="field"
                            >
                              <Input
                                style={
                                  Object {
                                    "marginTop": 25,
                                  }
                                }
                                type="text"
                              >
                                <div
                                  className="ui input"
                                  style={
                                    Object {
                                      "marginTop": 25,
                                    }
                                  }
                                >
                                  <Button
                                    as="button"
                                    color="green"
                                    key=".0"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    <Ref
                                      innerRef={
                                        Object {
                                          "current": <button
                                            class="ui green button"
                                            type="button"
                                          >
                                            Add
                                          </button>,
                                        }
                                      }
                                    >
                                      <RefFindNode
                                        innerRef={
                                          Object {
                                            "current": <button
                                              class="ui green button"
                                              type="button"
                                            >
                                              Add
                                            </button>,
                                          }
                                        }
                                      >
                                        <button
                                          className="ui green button"
                                          onClick={[Function]}
                                          type="button"
                                        >
                                          Add
                                        </button>
                                      </RefFindNode>
                                    </Ref>
                                  </Button>
                                </div>
                              </Input>
                            </div>
                          </FormField>
                        </FormInput>
                      </div>
                    </GridColumn>
                    <GridColumn
                      key="addDynamicFormInputs00/.0"
                    >
                      <div
                        className="column"
                      >
                        <FormInput
                          as={[Function]}
                          control={[Function]}
                          error={false}
                          label="First Name"
                          name="firstName"
                          onChange={[Function]}
                          value=""
                        >
                          <FormField
                            control={[Function]}
                            error={false}
                            label="First Name"
                            name="firstName"
                            onChange={[Function]}
                            value=""
                          >
                            <div
                              className="field"
                            >
                              <label>
                                First Name
                              </label>
                              <Input
                                name="firstName"
                                onChange={[Function]}
                                type="text"
                                value=""
                              >
                                <div
                                  className="ui input"
                                >
                                  <input
                                    name="firstName"
                                    onChange={[Function]}
                                    type="text"
                                    value=""
                                  />
                                </div>
                              </Input>
                            </div>
                          </FormField>
                        </FormInput>
                      </div>
                    </GridColumn>
                    <GridColumn
                      key="addDynamicFormInputs10/.1"
                    >
                      <div
                        className="column"
                      >
                        <FormInput
                          as={[Function]}
                          control={[Function]}
                          error={false}
                          label="Last Name"
                          name="lastName"
                          onChange={[Function]}
                          value=""
                        >
                          <FormField
                            control={[Function]}
                            error={false}
                            label="Last Name"
                            name="lastName"
                            onChange={[Function]}
                            value=""
                          >
                            <div
                              className="field"
                            >
                              <label>
                                Last Name
                              </label>
                              <Input
                                name="lastName"
                                onChange={[Function]}
                                type="text"
                                value=""
                              >
                                <div
                                  className="ui input"
                                >
                                  <input
                                    name="lastName"
                                    onChange={[Function]}
                                    type="text"
                                    value=""
                                  />
                                </div>
                              </Input>
                            </div>
                          </FormField>
                        </FormInput>
                      </div>
                    </GridColumn>
                  </div>
                </GridRow>
              </div>
            </Grid>
          </FieldArrayInner>
        </FormikConnect(FieldArrayInner)>
      </AddDynamicFormInput>
    </RenderForm>
  </Formik>
</TForm>
`;
